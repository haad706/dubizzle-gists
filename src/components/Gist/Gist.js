import styled from "styled-components";
import PropTypes from "prop-types";
import { Date, Files, IconText, UserTitle } from "components";

export const Gist = ({
  files = {},
  owner = {},
  forks_url: forksUrl,
  comments,
  comments_url: commentsUrl,
  created_at: createdAt,
  updated_at: updatedAt,
  description,
}) => {
  const { login, avatar_url: avatarImageUrl } = owner;

  const numberOfFiles = Object.keys(files).length;

  return (
    <CardWrapper>
      <HeaderWrapper>
        <UserTitle name={login} imageSrc={avatarImageUrl} />
        <HeaderInfoWrapper>
          <IconText iconName="code" text={`${numberOfFiles} Files`} />
          <IconText iconName="repo-forked" text="Forks" link={forksUrl} />
          <IconText
            iconName="comment"
            text={`${comments} Comments`}
            link={commentsUrl}
          />
          <IconText iconName="star" text="Stars" />
        </HeaderInfoWrapper>
      </HeaderWrapper>
      <DateWrapper>
        <Date label="Created at" time={createdAt} />
        <Date label="Updated at" time={updatedAt} />
      </DateWrapper>
      {description && <h4>{description}</h4>}
      <Files files={Object.values(files)} />
    </CardWrapper>
  );
};

const CardWrapper = styled.div`
  box-shadow: 0px 2px 16px rgba(0, 0, 0, 0.15);
  padding: 1.5rem;
  margin-top: 0.5rem;
  margin-bottom: 0.5rem;
  display: flex;
  flex-direction: column;
  gap: 1rem;
  border-radius: 10px;
  h4 {
    margin: 0;
    overflow: hidden;
    text-overflow: ellipsis;
    display: -webkit-box;
    -webkit-line-clamp: 2;
    -webkit-box-orient: vertical;
  }
`;

const HeaderWrapper = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
`;

const HeaderInfoWrapper = styled.div`
  display: flex;
  gap: 1rem;
  align-items: center;
`;

const DateWrapper = styled.div`
  display: flex;
  gap: 1rem;
  align-items: center;
`;

Gist.propTypes = {
  files: PropTypes.arrayOf(
    PropTypes.shape({
      filename: PropTypes.string.isRequired,
      raw_url: PropTypes.string.isRequired,
    })
  ),
  owner: PropTypes.shape({
    login: PropTypes.string.isRequired,
    avatarImageUrl: PropTypes.string.isRequired,
  }),
  forksUrl: PropTypes.string.isRequired,
  createdAt: PropTypes.string.isRequired,
  updatedAt: PropTypes.string.isRequired,
  commentsUrl: PropTypes.string.isRequired,
  comments: PropTypes.number.isRequired,
  description: PropTypes.string.isRequired,
};

export default Gist;
