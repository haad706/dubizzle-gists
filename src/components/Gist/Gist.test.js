import { render } from "@testing-library/react";
import Gist from "./Gist";

const gist = {
  files: {
    "gistfile1.txt": {
      filename: "gistfile1.txt",
      raw_url:
        "https://gist.githubusercontent.com/dwfault/80059b65e85973dbd629dff8d041588d/raw/3599c7641b180f2a4b6041d5e0f2e38f073f0fe6/gistfile1.txt",
    },
    "test.xml": {
      filename: "test.xml",
      raw_url:
        "https://gist.githubusercontent.com/Bezarius/cf8d6d227abb4450fa548c15148d5285/raw/b8aad7d46618739da077c0e0a84d26f805f2f929/test.xml",
    },
  },
  forks_url:
    "https://api.github.com/gists/3d62c5bac006167e2bf09b938e172bef/forks",
  owner: {
    avatar_url: "https://avatars.githubusercontent.com/u/39902740?v=4",
    login: "JianHongHong",
  },
  comments: 0,
  comments_url:
    "https://api.github.com/gists/3d62c5bac006167e2bf09b938e172bef/comments",
  created_at: "2023-05-24T06:51:40Z",
  updated_at: "2023-05-24T06:51:40Z",
  description:
    "Solution to level 1 in Untrusted: http://alex.nisnevich.com/untrusted/",
};

describe("Gist", () => {
  it("should render successfully", () => {
    const { baseElement } = render(
      <Gist {...gist} files={Object.values(gist.files)} />
    );
    expect(baseElement).toBeTruthy();
  });
  it("should match snapshot", () => {
    const { asFragment } = render(
      <Gist {...gist} files={Object.values(gist.files)} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
