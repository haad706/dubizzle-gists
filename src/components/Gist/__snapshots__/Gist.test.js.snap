// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gist should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-crozmw fTSxVO"
  >
    <div
      class="sc-jJTsDX cHXHtq"
    >
      <div
        class="sc-kPTPQs lhuErM"
      >
        <img
          class="sc-ispOId hkyLhQ"
          src="https://avatars.githubusercontent.com/u/39902740?v=4"
        />
        <p>
          JianHongHong
        </p>
      </div>
      <div
        class="sc-cJPUEC ePePQH"
      >
        <div
          class="sc-iVCKna jBWbRE"
        >
          <span
            class="octicon octicon-code"
          />
          <p>
            2 Files
          </p>
        </div>
        <div
          class="sc-iVCKna jBWbRE"
        >
          <span
            class="octicon octicon-repo-forked"
          />
          <a
            class="sc-gxYJeL fHpJfG"
            href="https://api.github.com/gists/3d62c5bac006167e2bf09b938e172bef/forks"
            target="_blank"
          >
            Forks
          </a>
        </div>
        <div
          class="sc-iVCKna jBWbRE"
        >
          <span
            class="octicon octicon-comment"
          />
          <a
            class="sc-gxYJeL fHpJfG"
            href="https://api.github.com/gists/3d62c5bac006167e2bf09b938e172bef/comments"
            target="_blank"
          >
            0 Comments
          </a>
        </div>
        <div
          class="sc-iVCKna jBWbRE"
        >
          <span
            class="octicon octicon-star"
          />
          <p>
            Stars
          </p>
        </div>
      </div>
    </div>
    <div
      class="sc-jeCNp hKVzdv"
    >
      <div
        class="sc-bgqQcB fppZgW"
      >
        <p>
          Created at:
        </p>
        <p>
          24/05/2023
        </p>
      </div>
      <div
        class="sc-bgqQcB fppZgW"
      >
        <p>
          Updated at:
        </p>
        <p>
          24/05/2023
        </p>
      </div>
    </div>
    <h4>
      Solution to level 1 in Untrusted: http://alex.nisnevich.com/untrusted/
    </h4>
    <div
      class="sc-ewnqHT pzFJb"
    >
      <div
        class="sc-iVCKna jBWbRE"
      >
        <span
          class="octicon octicon-file"
        />
        <a
          class="sc-gxYJeL fHpJfG"
          href="https://gist.githubusercontent.com/dwfault/80059b65e85973dbd629dff8d041588d/raw/3599c7641b180f2a4b6041d5e0f2e38f073f0fe6/gistfile1.txt"
          target="_blank"
        >
          gistfile1.txt
        </a>
      </div>
      <div
        class="sc-iVCKna jBWbRE"
      >
        <span
          class="octicon octicon-file"
        />
        <a
          class="sc-gxYJeL fHpJfG"
          href="https://gist.githubusercontent.com/Bezarius/cf8d6d227abb4450fa548c15148d5285/raw/b8aad7d46618739da077c0e0a84d26f805f2f929/test.xml"
          target="_blank"
        >
          test.xml
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
